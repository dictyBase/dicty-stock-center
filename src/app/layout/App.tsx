import React, { useCallback, useRef } from "react"
import { useQuery } from "@apollo/client"
import { Header, Footer } from "dicty-components-header-footer"
import { Navbar } from "dicty-components-navbar"
import jwtDecode from "jwt-decode"
import { CartProvider } from "features/ShoppingCart/CartStore"
import { IconProp } from "@fortawesome/fontawesome-svg-core"
import { useFetchRefreshToken, useFooter, useNavbar } from "dicty-hooks"
import {
  headerItems,
  loggedHeaderItems,
  generateLinks,
} from "common/utils/headerItems"
import CartIcon from "features/ShoppingCart/CartIcon"
import ErrorBoundary from "features/Errors/ErrorBoundary"
import RenderRoutes from "app/routes/RenderRoutes"
import { useAuthStore, ActionType } from "features/Authentication/AuthStore"
import { GET_REFRESH_TOKEN } from "common/graphql/queries/auth"
import { User } from "common/@types/user"
import { useStyles, navTheme } from "./appStyles"

type HeaderProps = {
  /** Identifies if header link should use React Router */
  isRouter?: boolean
  /** Text to display in link */
  text: string
  /** Font Awesome Icon to display */
  icon: IconProp
  /** URL of link */
  url: string
}

const getTokenIntervalDelayInMS = (token: string) => {
  if (token === "") {
    return
  }
  const decodedToken = jwtDecode(token) as any
  const currentTime = new Date(Date.now())
  const jwtTime = new Date(decodedToken.exp * 1000)
  const timeDiffInMins = (+jwtTime - +currentTime) / 60000
  // all this to say we want the delay to be two minutes before the JWT expires
  return (timeDiffInMins - 2) * 60 * 1000
}

type RefreshTokenData = {
  token: string
  user: User
  identity: {
    provider: string
  }
}

type Action = {
  type: string
  payload: {
    provider: string
    token: string
    user: User
  }
}

const updateToken = (
  dispatch: (arg0: Action) => void,
  data: RefreshTokenData,
) =>
  dispatch({
    type: ActionType.UPDATE_TOKEN,
    payload: {
      provider: data.identity.provider,
      token: data.token,
      user: data.user,
    },
  })

const App = () => {
  const [skip, setSkip] = React.useState(false)
  const [{ isAuthenticated, token }, dispatch] = useAuthStore()
  const { navbarData } = useNavbar()
  const { footerData } = useFooter()
  const classes = useStyles()
  const { loading, refetch, data } = useQuery(GET_REFRESH_TOKEN, {
    variables: { token: token },
    errorPolicy: "ignore",
    skip, // only run query once
  })
  const interval = useRef<any>(null)
  const delay = getTokenIntervalDelayInMS(token)

  // set skip to true so the query is only run once
  // then update the refresh token in our global state
  // https://github.com/trojanowski/react-apollo-hooks/issues/157#issuecomment-491151909
  React.useEffect(() => {
    if (!loading && data && data.getRefreshToken) {
      setSkip(true)
      updateToken(dispatch, data.getRefreshToken)
    }
  }, [data, dispatch, loading])

  const fetchRefreshToken = useCallback(async () => {
    try {
      const res = await refetch({ token: token })
      if (res.data.getRefreshToken) {
        const { data } = res
        updateToken(dispatch, data.getRefreshToken)
      }
    } catch (error) {
      console.error(error)
    }
  }, [dispatch, refetch, token])
  useFetchRefreshToken(fetchRefreshToken, interval, delay!, isAuthenticated)

  const headerContent = isAuthenticated ? loggedHeaderItems : headerItems

  return (
    <div className={classes.body}>
      <Header items={headerContent}>
        {(items: Array<HeaderProps>) => items.map(generateLinks)}
      </Header>
      <Navbar items={navbarData} theme={navTheme} />
      <br />
      <CartProvider>
        <main className={classes.main}>
          <CartIcon />
          <ErrorBoundary>
            <RenderRoutes />
          </ErrorBoundary>
        </main>
      </CartProvider>
      <Footer items={footerData} />
    </div>
  )
}

export { getTokenIntervalDelayInMS }
export default App
