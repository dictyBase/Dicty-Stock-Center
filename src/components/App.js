// @flow
import React, { useEffect } from "react"
import { withRouter } from "react-router-dom"
import { connect } from "react-redux"
import { Header, Footer } from "dicty-components-header-footer"
import { Navbar } from "dicty-components-navbar"
import fetchNavbarAndFooter from "actions/navbar"
import footerItems from "constants/Footer"
import navItems from "constants/Navbar"
import { CartProvider } from "components/ShoppingCart/CartStore"
import {
  headerItems,
  loggedHeaderItems,
  generateLinks,
} from "utils/headerItems"
import "utils/icons" // fontawesome library
import CartIcon from "components/ShoppingCart/CartIcon"
import ErrorBoundary from "components/Errors/ErrorBoundary"
import RenderRoutes from "routes/RenderRoutes"
import { useStyles, navTheme } from "./appStyles"

type Props = {
  /** Object representing auth part of state */
  auth: Object,
  /** Object representing navbar part of state */
  navbar: Object,
  /** Object representing footer part of state */
  footer: Object,
  /** Action creator to fetch navbar and footer content */
  fetchNavbarAndFooter: Function,
}

const App = (props: Props) => {
  const { auth, navbar, footer, fetchNavbarAndFooter } = props
  const classes = useStyles()
  const headerContent = auth.isAuthenticated ? loggedHeaderItems : headerItems
  // if any errors, fall back to old link setup
  const navbarContent = !navbar.links ? navItems : navbar.links
  const footerContent = !footer.links ? footerItems : footer.links

  useEffect(() => {
    fetchNavbarAndFooter()
  }, [fetchNavbarAndFooter])

  return (
    <div className={classes.body}>
      <Header items={headerContent}>{items => items.map(generateLinks)}</Header>
      <Navbar items={navbarContent} theme={navTheme} />
      <br />
      <CartProvider>
        <main className={classes.main}>
          <CartIcon />
          <ErrorBoundary>
            <RenderRoutes {...props} />
          </ErrorBoundary>
        </main>
      </CartProvider>
      <Footer items={footerContent} />
    </div>
  )
}

const mapStateToProps = ({ auth, navbar, footer }) => ({
  auth,
  navbar,
  footer,
})

export { App }
export default withRouter<*, *>(
  connect<*, *, *, *, *, *>(mapStateToProps, { fetchNavbarAndFooter })(App),
)
