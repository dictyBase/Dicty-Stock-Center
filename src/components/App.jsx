// @flow
import React from "react"
import { Header, Footer } from "dicty-components-header-footer"
import { Navbar } from "dicty-components-navbar"
import {
  headerItems,
  loggedHeaderItems,
  generateLinks,
} from "utils/headerItems"
import Cart from "components/Cart"
import RenderRoutes from "routes/RenderRoutes"
import { withRouter } from "react-router-dom"
import { connect } from "react-redux"
import { FooterLinks } from "constants/Rooter"
import { NavbarLinks } from "constants/Navbar"
import type { MapStateToProps } from "react-redux"

type Props = {
  cart: Object,
  auth: Object,
}

const App = (props: Props) => {
  return (
    <div>
      {props.auth.isAuthenticated ? (
        <Header items={loggedHeaderItems}>
          {items => items.map(generateLinks)}
        </Header>
      ) : (
        <Header items={headerItems}>{items => items.map(generateLinks)}</Header>
      )}
      <br />
      <Navbar items={NavbarLinks} />
      <br />
      <Cart cart={props.cart} />
      <RenderRoutes {...props} />
      <Footer items={FooterLinks} />
    </div>
  )
}

const mapStateToProps: MapStateToProps<*, *, *> = ({ auth, cart }) => ({
  auth,
  cart,
})

export default withRouter(connect(mapStateToProps)(App))
