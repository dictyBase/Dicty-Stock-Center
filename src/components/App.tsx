import React, { useCallback, useRef } from "react"
import { useQuery } from "@apollo/react-hooks"
import { Header, Footer } from "dicty-components-header-footer"
import { Navbar } from "dicty-components-navbar"
import jwtDecode from "jwt-decode"
import { CartProvider } from "components/ShoppingCart/CartStore"
import { IconProp } from "@fortawesome/fontawesome-svg-core"
import {
  headerItems,
  loggedHeaderItems,
  generateLinks,
} from "utils/headerItems"
import CartIcon from "components/ShoppingCart/CartIcon"
import ErrorBoundary from "components/Errors/ErrorBoundary"
import RenderRoutes from "routes/RenderRoutes"
import { useAuthStore } from "components/authentication/AuthStore"
import { useFetchRefreshToken, useFooter, useNavbar } from "dicty-hooks"
import { GET_REFRESH_TOKEN } from "graphql/queries"
import { useStyles, navTheme } from "./appStyles"

interface IHeader {
  isRouter?: boolean
  text: string
  icon: IconProp
  url: string
}

const getTokenIntervalDelayInMS = (token: string) => {
  if (token === "") {
    return
  }
  const decodedToken = jwtDecode(token) as any
  const currentTime = new Date(Date.now())
  const jwtTime = new Date(decodedToken.exp * 1000)
  const timeDiffInMins = (+jwtTime - +currentTime) / 60000
  // all this to say we want the delay to be two minutes before the JWT expires
  return (timeDiffInMins - 2) * 60 * 1000
}

const App = () => {
  const [{ isAuthenticated, token }, dispatch] = useAuthStore()
  const { navbarData } = useNavbar()
  const { footerData } = useFooter()
  const classes = useStyles()
  const { refetch } = useQuery(GET_REFRESH_TOKEN, {
    variables: { token: token },
    errorPolicy: "ignore",
  })
  const interval = useRef(null)
  const delay = getTokenIntervalDelayInMS(token)

  const fetchRefreshToken = useCallback(async () => {
    const res = await refetch({ token: token })
    if (res.data.getRefreshToken) {
      const { data } = res
      dispatch({
        type: "UPDATE_TOKEN",
        payload: {
          provider: data.getRefreshToken.identity.provider,
          token: data.getRefreshToken.token,
          user: data.getRefreshToken.user,
        },
      })
    }
  }, [dispatch, refetch, token])
  useFetchRefreshToken(fetchRefreshToken, interval, delay!, isAuthenticated)

  const headerContent = isAuthenticated ? loggedHeaderItems : headerItems

  return (
    <div className={classes.body}>
      <Header items={headerContent}>
        {(items: Array<IHeader>) => items.map(generateLinks)}
      </Header>
      <Navbar items={navbarData} theme={navTheme} />
      <br />
      <CartProvider>
        <main className={classes.main}>
          <CartIcon />
          <ErrorBoundary>
            <RenderRoutes />
          </ErrorBoundary>
        </main>
      </CartProvider>
      <Footer items={footerData} />
    </div>
  )
}

export { getTokenIntervalDelayInMS }
export default App
