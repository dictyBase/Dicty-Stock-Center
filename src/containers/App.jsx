import React, { Component, PropTypes } from 'react'
import { Header, Footer } from 'dicty-components-header-footer'
import { Navbar } from 'dicty-components-navbar'
import Cart from 'components/Cart'
import Routes from 'components/Routes'
import { connect } from 'react-redux'
import { StyleRoot } from 'radium'
import { bindActionCreators } from 'redux'
import * as authActionCreators from 'actions/auth'
import * as shippingActionCreators from 'actions/order/shipping'
import * as paymentActionCreators from 'actions/order/payment'
import * as submitActionCreators from 'actions/order/submit'
import * as pageActionCreators from 'actions/page'
import * as dscActionsCreators from 'actions/stockCenter'
import * as cartActionCreators from 'actions/cart'
import { routerActions } from 'react-router-redux'
import { FooterLinks } from 'constants/Footer'
import { NavbarLinks } from 'constants/Navbar'

class App extends Component {
    displayName = 'the primary app component';
    static propTypes = {
        auth: PropTypes.object.isRequired,
        authActions: PropTypes.object.isRequired,
        stockCenter: PropTypes.object.isRequired,
        stockCenterActions: PropTypes.object.isRequired
    };
    renderChildren = () => {
        const { children } = this.props
        return React.Children.map(children, (child) => {
            return React.cloneElement(child, {...this.props})
        })
    };
    render() {
        return (
            <StyleRoot>
                <div>
                    <Header
                      auth={ this.props.auth }
                      authActions={ this.props.authActions }
                      downloads=""
                      info=""
                      cite=""
                    />
                    <Navbar items={ NavbarLinks } />
                    <Cart cart={ this.props.cart }/>
                    <Routes />
                    <Footer items={ FooterLinks } />
                </div>
            </StyleRoot>
        )
    }
}

const mapStateToProps = (state, ownProps) => {
    const { auth, order, page, stockCenter, cart } = state
    return {
        auth: auth,
        routeProps: ownProps,
        order: order,
        page: page,
        stockCenter: stockCenter,
        cart: cart
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        authActions: bindActionCreators(authActionCreators, dispatch),
        orderActions: bindActionCreators(
            Object.assign({}, shippingActionCreators, paymentActionCreators, submitActionCreators),
            dispatch
        ),
        pageActions: bindActionCreators(pageActionCreators, dispatch),
        routerActions: bindActionCreators(routerActions, dispatch),
        stockCenterActions: bindActionCreators(dscActionsCreators, dispatch),
        cartActions: bindActionCreators(cartActionCreators, dispatch)
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(App)
