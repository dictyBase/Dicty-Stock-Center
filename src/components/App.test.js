import React from "react"
import toJson from "enzyme-to-json"
import { shallow } from "enzyme"
import "jest-styled-components"
import { App } from "./App"

test("matching a snapshot of App", () => {
  const props = {
    auth: {
      isAuthenticated: false,
    },
    cart: {},
    footer: {},
    navbar: {},
    fetchNavbarAction: () => {},
    fetchFooterAction: () => {},
  }
  const subject = shallow(<App {...props} />)
  expect(toJson(subject)).toMatchSnapshot()
})
