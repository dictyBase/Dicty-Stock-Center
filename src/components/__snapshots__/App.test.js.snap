// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matching a snapshot of App 1`] = `
<div>
  <WithStyles(Header)
    items={
      Array [
        Object {
          "icon": "plus",
          "text": "Cite Us",
          "url": "/cite",
        },
        Object {
          "icon": "download",
          "text": "Downloads",
          "url": "/downloads",
        },
        Object {
          "icon": "info-circle",
          "text": "About dictyBase",
          "url": "/info",
        },
        Object {
          "icon": "sign-in",
          "isRouter": true,
          "text": "Login",
          "url": "/login",
        },
      ]
    }
  />
  <br />
  <Navbar
    items={
      Array [
        Object {
          "dropdown": true,
          "items": Array [
            Object {
              "href": "/",
              "name": "Dictyostelium discoideum",
            },
            Object {
              "href": "http://genomes.dictybase.org/purpureum",
              "name": "Dictyostelium purpureum",
            },
            Object {
              "href": "http://genomes.dictybase.org/fasciculatum",
              "name": "Dictyostelium fasciculatum",
            },
            Object {
              "href": "http://genomes.dictybase.org/pallidum",
              "name": "Polysphondylium pallium",
            },
          ],
          "title": "Genomes",
        },
        Object {
          "dropdown": true,
          "items": Array [
            Object {
              "href": "http://dictybase.org/tools/jbrowse/?data=data/jbrowse/discoideum&loc=6:1..50011&tracks=reference,gene,transcript",
              "name": "Genome Browser",
            },
            Object {
              "href": "/dictyaccess",
              "name": "Dashboard",
            },
          ],
          "title": "Tools",
        },
        Object {
          "dropdown": true,
          "items": Array [
            Object {
              "href": "/explore/art",
              "name": "Dicty Art",
            },
            Object {
              "href": "/explore/gallery",
              "name": "Gallery",
            },
            Object {
              "href": "/explore/learn",
              "name": "Learn About Dicty",
            },
            Object {
              "href": "/explore/teach",
              "name": "Teaching Protocols",
            },
            Object {
              "href": "/explore/links",
              "name": "Useful Links",
            },
          ],
          "title": "Explore",
        },
        Object {
          "dropdown": true,
          "items": Array [
            Object {
              "href": "/research/ontology",
              "name": "Anatomy Ontology",
            },
            Object {
              "href": "/research/codon",
              "name": "Codon Bias Table",
            },
            Object {
              "href": "/research/nomenclature",
              "name": "Nomenclature Guidelines",
            },
            Object {
              "href": "/research/phenotyping",
              "name": "Phenotyping",
            },
            Object {
              "href": "/research/techniques",
              "name": "Techniques",
            },
          ],
          "title": "Research",
        },
        Object {
          "dropdown": true,
          "items": Array [
            Object {
              "href": "/stockcenter",
              "name": "Stock Center Home",
            },
            Object {
              "href": "/stockcenter/information/order",
              "name": "Order Information",
            },
            Object {
              "href": "/stockcenter/information/faq",
              "name": "FAQ",
            },
          ],
          "title": "Dicty Stock Center",
        },
        Object {
          "dropdown": true,
          "items": Array [
            Object {
              "href": "/community/citation",
              "name": "Cite Us",
            },
            Object {
              "href": "/community/conference",
              "name": "Dicty Annual Conferences",
            },
            Object {
              "href": "/community/listserv",
              "name": "Dicty Email Forum",
            },
            Object {
              "href": "/community/labs",
              "name": "Dicty Labs",
            },
            Object {
              "href": "/community/history",
              "name": "History",
            },
            Object {
              "href": "/community/jobs",
              "name": "Jobs",
            },
            Object {
              "href": "/community/meetings",
              "name": "Upcoming Meetings",
            },
          ],
          "title": "Community",
        },
      ]
    }
  />
  <br />
  <Connect(Component)
    cart={Object {}}
  />
  <styled.div>
    <RenderRoutes
      auth={
        Object {
          "isAuthenticated": false,
        }
      }
      cart={Object {}}
      fetchFooterAction={[Function]}
      fetchNavbarAction={[Function]}
      footer={Object {}}
      navbar={Object {}}
    />
  </styled.div>
  <WithStyles(Footer)
    items={
      Array [
        Array [
          Object {
            "header": Object {
              "description": "Genomes",
            },
            "items": Array [
              Object {
                "description": "Dictyostelium discoideum",
                "link": "/",
              },
              Object {
                "description": "Dictyostelium purpureum",
                "link": "http://genomes.dictybase.org/purpureum",
              },
              Object {
                "description": "Dictyostelium fasciculatum",
                "link": "http://genomes.dictybase.org/fasciculatum",
              },
              Object {
                "description": "Polysphondylium pallium",
                "link": "http://genomes.dictybase.org/pallidum",
              },
            ],
          },
        ],
        Array [
          Object {
            "header": Object {
              "description": "Tools",
            },
            "items": Array [
              Object {
                "description": "Genome Browser",
                "link": "http://dictybase.org/tools/jbrowse/?data=data/jbrowse/discoideum&loc=6:1..50011&tracks=reference,gene,transcript",
              },
              Object {
                "description": "Dashboard",
                "link": "/dictyaccess",
              },
            ],
          },
        ],
        Array [
          Object {
            "header": Object {
              "description": "Explore",
            },
            "items": Array [
              Object {
                "description": "Dicty Art",
                "link": "/explore/art",
              },
              Object {
                "description": "Gallery",
                "link": "/explore/gallery",
              },
              Object {
                "description": "Learn About Dicty",
                "link": "/explore/learn",
              },
              Object {
                "description": "Teaching Protocols",
                "link": "/explore/teach",
              },
              Object {
                "description": "Useful Links",
                "link": "/explore/links",
              },
            ],
          },
        ],
        Array [
          Object {
            "header": Object {
              "description": "Research",
            },
            "items": Array [
              Object {
                "description": "Anatomy Ontology",
                "link": "/research/ontology",
              },
              Object {
                "description": "Codon Bias Table",
                "link": "/research/codon",
              },
              Object {
                "description": "Nomenclature Guidelines",
                "link": "/research/nomenclature",
              },
              Object {
                "description": "Phenotyping",
                "link": "/research/phenotyping",
              },
              Object {
                "description": "Techniques",
                "link": "/research/techniques",
              },
            ],
          },
        ],
        Array [
          Object {
            "header": Object {
              "description": "Dicty Stock Center",
            },
            "items": Array [
              Object {
                "description": "Stock Center Home",
                "link": "/stockcenter",
              },
              Object {
                "description": "Order Information",
                "link": "/stockcenter/information/order",
              },
              Object {
                "description": "FAQ",
                "link": "/stockcenter/information/faq",
              },
            ],
          },
        ],
        Array [
          Object {
            "header": Object {
              "description": "Community",
            },
            "items": Array [
              Object {
                "description": "Cite Us",
                "link": "/community/citation",
              },
              Object {
                "description": "Dicty Annual Conferences",
                "link": "/community/conference",
              },
              Object {
                "description": "Dicty Email Forum",
                "link": "/community/listserv",
              },
              Object {
                "description": "Dicty Labs",
                "link": "/community/labs",
              },
              Object {
                "description": "History",
                "link": "/community/history",
              },
              Object {
                "description": "Jobs",
                "link": "/community/jobs",
              },
              Object {
                "description": "Upcoming Meetings",
                "link": "/community/meetings",
              },
            ],
          },
        ],
        Array [
          Object {
            "header": Object {
              "description": "Please Cite",
            },
            "items": Array [
              Object {
                "description": "dictyBase",
                "link": "",
              },
              Object {
                "description": "Dicty Stock Center",
                "link": "/stockcenter",
              },
            ],
          },
          Object {
            "header": Object {
              "description": "Supported By",
            },
            "items": Array [
              Object {
                "description": "NIH",
                "link": "https://www.nih.gov/",
              },
              Object {
                "description": "GMOD",
                "link": "http://gmod.org/wiki/Main_Page",
              },
              Object {
                "description": "Gene Ontology",
                "link": "http://www.geneontology.org/",
              },
            ],
          },
        ],
      ]
    }
  />
</div>
`;
